File: frontend/packages/shared/src/types/session.ts
import type { WorktreeRef } from "./worktree";

/**
 * Approval mode for a session (kernel V5+1aae940; live-reconcile since
 * V5+bba3014). ``default`` parks every write / shell / MCP call on the
 * host for explicit user approval. ``auto_review`` lets an LLM
 * classifier auto-decide and surfaces a review feed (Claude tier only —
 * the API rejects this for DeepAgents). ``full_access`` auto-approves
 * everything (current default, matches the legacy "bypass" role).
 *
 * Mutable mid-session via ``PATCH /v1/sessions/{id}/permission-mode``;
 * applies on the next Send.
 */
export type SessionPermissionMode = "default" | "auto_review" | "full_access";

/**
 * Cross-runtime reasoning-effort budget (kernel V5+bba3014
 * ``ModelSettings.effort``). Mutable mid-session via
 * ``PATCH /v1/sessions/{id}/effort``; applies on next Send.
 *
 * Per-runtime value mapping:
 *   * Claude: pass-through all 5 values.
 *   * Codex: ``max`` clamps to ``xhigh``.
 *   * DeepAgents openai-completion: ``max`` → ``xhigh``.
 *   * DeepAgents anthropic: pass-through.
 *   * DeepAgents gemini: ``xhigh|max`` → ``high``.
 */
export type EffortLevel = "low" | "medium" | "high" | "xhigh" | "max";

/**
 * Session working mode (kernel ``Session.mode``,
 * docs/design/session-modes.md). ``plan`` = the runtime plans before
 * touching anything (Claude lowers to SDK ``permissionMode="plan"`` +
 * the ``ExitPlanMode`` approval card); ``goal`` = the runtime loops
 * until a goal condition is met (task lead/member sessions). Only
 * ``claude_agent`` / ``codex`` sessions accept non-default modes.
 *
 * Mutable mid-session via ``PATCH /v1/sessions/{id}/mode``. The runtime
 * can also exit a mode on its own (approved plan / completed goal) —
 * surfaced live as a ``session.mode_changed`` SSE frame
 * (``{mode, by: "user" | "runtime"}``).
 */
export type SessionMode = "default" | "plan" | "goal";

/**
 * Runtimes whose plan-mode lowering is wired end-to-end today.
 * ``claude_agent`` lowers natively (typed SDK
 * ``set_permission_mode("plan")`` + the ``ExitPlanMode`` approval
 * card); ``codex`` via ``turn/start.collaborationMode`` (plan proposal
 * arrives as ``session.plan_proposed`` and approval is client-driven).
 * deepagents / deepseek_harness have no native primitive and the
 * server 400s them. UI affordances gate on this list so users never
 * see a toggle the backend would reject.
 */
export const PLAN_MODE_RUNTIMES: readonly string[] = ["claude_agent", "codex"];

/** Whether the composer should offer the plan-mode toggle for a runtime. */
export function supportsPlanMode(
  runtimeProvider: string | null | undefined,
): boolean {
  return runtimeProvider != null && PLAN_MODE_RUNTIMES.includes(runtimeProvider);
}

export interface SessionListItem {
  id: string;
  project_id: string;
  name: string | null;
  status: "created" | "running" | "idle" | "failed" | "cancelled" | "archived";
  origin: "user" | "schedule" | "automation";
  /** CLIENT-side tag on multi-target editions: which execution target
   * answered the list row (e.g. "local"/"cloud"). Distinct from ``origin``
   * (the server-side initiator). Never sent by the server. */
  exec_origin?: string;
  last_user_message_text: string | null;
  locked_model_id: string | null;
  /**
   * Provider id stamped at session creation. Promoted from SessionDetail to
   * the list shape because the desktop sidebar's session list feeds the
   * composer's model selector — the selector matches on (providerId, modelId)
   * pairs, so without provider_id the locked model can't be displayed and
   * falls back to the project default provider.
   */
  locked_provider_id: string | null;
  runtime_provider: string;
  /**
   * Approval mode for this session. Live-reconcilable (kernel
   * V5+bba3014). Surfaced on the list shape so the UI can render an
   * "auto" / "needs review" badge without a second detail fetch.
   */
  permission_mode: SessionPermissionMode;
  /**
   * Reasoning-effort budget for this session (kernel V5+bba3014
   * ``ModelSettings.effort``). ``null`` = SDK default. Live-
   * reconcilable. Surfaced on the list shape so the composer can
   * render the EffortSelector's current value without a second fetch.
   */
  effort: EffortLevel | null;
  /**
   * Session working mode. Surfaced on the list shape so the composer's
   * Plan chip reflects the session's current mode without a second
   * fetch. Optional because older backends omit it — treat a missing
   * value as ``"default"``.
   */
  mode?: SessionMode;
  /**
   * Owning task id if this session belongs to a task run (lead session
   * or dispatched sub-Run). ``null`` = a user-initiated standalone
   * conversation. Read from ``session.metadata.valuz.task_id``. The
   * desktop sidebar's "recent" list filters on this so task-internal
   * sessions don't leak into the user's chat history.
   */
  task_id: string | null;
  /**
   * Worktree the session runs in (immutable creation-time snapshot).
   * `null` = the project's main workspace. On the *detail* shape the
   * backend also populates `worktree.exists` (liveness, computed on read).
   */
  worktree?: WorktreeRef | null;
  /**
   * Fork provenance (docs/design/session-fork.md): the source session this
   * one was forked from. `null` for sessions that are not forks. The source
   * may since have been deleted — a navigation hint, not a guarantee.
   */
  forked_from_session_id?: string | null;
  /** Unix epoch milliseconds (UTC). Format via `new Date(ms)`. */
  updated_at: number;
  /**
   * Aggregated token usage when the row originated from a session-detail
   * response. List endpoints may omit it to avoid an extra message scan.
   */
  total_tokens?: number;
  /**
   * A `run_in_background` task is still executing in this session. Same fact
   * and same source as `RunSummary.background` — both read the orchestrator's
   * live registry via `bg_busy_session_ids()` — so the conversation header,
   * the sidebar pulse and the Activity page can never disagree about one
   * session.
   *
   * Deliberately separate from `status`: the launching turn genuinely ends,
   * and `status` drives the Stop button and queue routing, so a `running`
   * value here would offer a Stop that stops nothing and route new messages
   * into the queue (409 "Session is already running").
   */
  background?: boolean;
}

/**
 * One row in the agent's TODO list snapshot. Mirrors the kernel's
 * `Session.todos` element shape, which itself matches the Claude Agent
 * SDK's TodoWrite payload verbatim — the camelCase `activeForm` is
 * preserved on the wire to avoid (de)serialise overhead and to keep the
 * shape identical to upstream.
 */
export interface TodoItem {
  content: string;
  /**
   * "pending" | "in_progress" | "completed". Typed as string for
   * forward-compat with future SDK additions.
   */
  status: string;
  /**
   * Gerund form ("Planning migration") rendered while in_progress.
   * Some runtimes (DeepAgents) omit it — fall back to `content`.
   */
  activeForm?: string | null;
}

export interface SessionDetail extends SessionListItem {
  total_tokens: number;
  total_cost_usd: number;
  /** Unix epoch milliseconds (UTC). Format via `new Date(ms)`. */
  created_at: number | null;
  /**
   * Free-form metadata captured at session creation. Currently used to
   * remember UI mode hints (e.g. {"mode": "skill-creator"}) so revisiting
   * the session from history restores the right panel/banner.
   */
  trigger_meta?: Record<string, string> | null;
  /**
   * Latest TODO snapshot from the kernel session, populated as the agent
   * calls TodoWrite. `null` means "no todos yet"; an empty array is a
   * meaningful "all done" signal preserved by the kernel.
   */
  todos?: TodoItem[] | null;
  /**
   * Frozen system-prompt append captured at session creation from the
   * project's then-current `instructions_md` (ADR-008). The runtime
   * hands this verbatim to the model on every turn; project edits
   * after creation do *not* mutate this field. `null` when the project
   * had no instructions at session-create time. Session panels should
   * render this — NOT the live project `instructions_md` — so users
   * see the prompt the running agent actually has.
   */
  instructions?: string | null;
  /**
   * Project-local agent handle when the session was created from a named
   * project agent (e.g. a Project Task lead/member). `null` for plain chat
   * sessions not bound to a named agent.
   */
  agent_slug?: string | null;
}

// ---------------------------------------------------------------------------
// Session event types (wire-level shapes shared by core + ui + apps).
// Moved from core/api/sessions-api.ts — pure data, no runtime deps.
// ---------------------------------------------------------------------------

export interface SessionEventDTO {
  /**
   * PER-STORE row id. The backend has TWO independent seq spaces: history
   * reads (REST list/window, SSE backfill frames, heartbeats) carry the
   * DURABLE store's seq, while live-stream frames carry the kernel's LOCAL
   * seq. Seqs from the two spaces must NEVER be compared against each other
   * or fed into one cursor — cross-store identity is ``event_uid``.
   * ``0`` marks a live unpersisted frame (streaming deltas).
   */
  seq: number;
  event: {
    event_type: string;
    payload: Record<string, string>;
  };
  /**
   * Unix epoch milliseconds (UTC) the kernel persisted the event with.
   * The SSE wire emits it (``SessionEventFrame.to_sse_data``); listEvents
   * mirrors it through ``SessionEventEnvelope.timestamp``. ``undefined``
   * for synthetic envelopes that don't have one. Format via `new Date(ms)`.
   */
  timestamp?: number;
  /**
   * Store-independent identity of a persisted event (32-hex string), present
   * on both history and live frames. This is the ONLY key valid for
   * cross-segment (history ↔ live) dedup/merge — ``seq`` is per-store.
   * ``null``/``undefined`` on live-only delta frames (never persisted) and
   * on legacy rows persisted before uid minting; those keep the historical
   * seq-based behavior.
   */
  event_uid?: string | null;
}

/**
 * Subjects the runtime can park on. ``shell_command`` / ``file_change``
 * / ``mcp_tool_call`` / ``tool_input`` all use the
 * ``approve | reject`` decision set; ``clarifying_questions`` uses the
 * ``answer | reject`` set (Claude SDK ``AskUserQuestion``). The host
 * preserves whatever string the kernel emits so future subjects flow
 * through transparently.
 */
export type RequiresActionSubject =
  | "shell_command"
  | "file_change"
  | "mcp_tool_call"
  | "tool_input"
  | "clarifying_questions"
  | (string & {});

/**
 * Preview of the session-scoped rule the kernel will commit if the
 * user picks "Always for this session" (kernel V5+d008b53). Each
 * runtime ships its own matcher so the ``display`` string reflects
 * the *native* rule grammar of the runtime (Claude pattern like
 * ``Bash(npm test:*)``, codex's MCP endpoint-level identity, exact
 * args for the default matcher). ``rule_data`` is the opaque payload
 * the matcher needs to match a future call — the frontend never
 * inspects it, only round-trips it through the rule preview UI.
 *
 * ``null`` for the ``clarifying_questions`` subject and for any
 * runtime that doesn't advertise ``approve_for_session`` in
 * ``available_decisions``.
 */
export interface SessionRulePreview {
  kind: string;
  display: string;
  runtime_kind: string;
  rule_data: Record<string, unknown>;
}

export interface RequiresActionEvent {
  message_id: string;
  pending_id: string;
  subject: RequiresActionSubject;
  runtime_provider: string;
  /**
   * Decisions the runtime advertised for this pending. v1 always
   * contains ``approve`` + ``reject``; v2 may add
   * ``approve_with_changes`` (Claude / DeepAgents) and
   * ``approve_for_session`` (every runtime that ships a rule matcher).
   * The clarifying_questions subject uses ``answer`` + ``reject``.
   */
  available_decisions: string[];
  /** Decoded subject-specific payload. Empty object on parse failure. */
  payload: Record<string, unknown>;
  /** ISO-8601 expiry; "" when the kernel didn't supply one. */
  expires_at: string;
  /**
   * V5+d008b53: structured preview of the session-scoped rule the
   * kernel will commit if the user picks "Always for this session".
   * ``null`` when the runtime doesn't advertise the verb (which is
   * also why ``approve_for_session`` won't appear in
   * ``available_decisions`` for that pending).
   */
  session_rule_preview: SessionRulePreview | null;
  /**
   * V5+d008b53: the original tool args the runtime parked on. The
   * frontend "Edit & Approve" JSON editor seeds from this when the
   * user opens the modal so they edit the *complete* args dict
   * rather than typing one from scratch. ``null`` for the
   * clarifying_questions subject (where args aren't tool args).
   */
  original_input: Record<string, unknown> | null;
}

export interface ActionResolvedEvent {
  message_id: string;
  pending_id: string;
  decision:
    | "approve"
    | "approve_with_changes"
    | "approve_for_session"
    | "auto_approved"
    | "reject"
    | "answer"
    | "expired"
    | "interrupted";
  resolved_by: "user" | "system" | (string & {});
  message: string;
  /** Decoded ``answers`` map, only populated when ``decision === "answer"``. */
  answers: Record<string, string | string[]>;
  /**
   * V5+d008b53: UUID of the committed session-scoped rule.
   * Non-null only when ``decision === "approve_for_session"``;
   * idempotent retries surface the original rule id.
   */
  rule_id: string | null;
  /**
   * V5+d008b53: rule that fired on this auto-approve.
   * Non-null only when ``decision === "auto_approved"``. The
   * frontend uses this to render the "auto-approved by rule X"
   * strip even though there was no preceding ``requires_action``.
   */
  auto_resolved_by_rule_id: string | null;
}

/**
 * One agent's slot in a Claude dynamic-workflow run. The kernel folds the
 * run's ``journal.jsonl`` into one entry per agent: ``progress`` while the
 * agent is running, ``done`` once its result lands. Extra fields (label /
 * phase) may appear in the terminal ``wf_<id>.json`` snapshot, which the
 * kernel forwards raw — so this type stays tolerant of unknown shapes.
 */
export interface WorkflowAgentProgress {
  /** Always ``"workflow_agent"`` for the live snapshot; tolerated as a free
   *  string so a richer terminal snapshot type flows through unchanged. */
  type?: string;
  agentId: string;
  state: "progress" | "done" | (string & {});
  /** Optional human label from the terminal snapshot (e.g. the agent's task). */
  label?: string;
  /** Optional phase grouping from the terminal snapshot. */
  phase?: string;
}

/**
 * A snapshot of a Claude dynamic-workflow (``Workflow`` tool) run, streamed
 * live by the kernel while the background runtime executes. The live snapshot
 * carries the fields below; the terminal ``wf_<id>.json`` result file is
 * forwarded raw and may carry extras (consumed defensively by the renderer).
 */
export interface WorkflowState {
  runId: string;
  workflowName: string | null;
  /** ``running`` during the run; ``completed`` / a terminal verb at the end. */
  status: "running" | "completed" | (string & {});
  agentCount: number;
  agentsDone: number;
  workflowProgress: WorkflowAgentProgress[];
  /** Path to the run's script on disk (first snapshot only). */
  scriptPath?: string;
  /** The workflow script content (first snapshot only); ``null`` if unread. */
  script?: string | null;
  /**
   * Path to the run's full result file (``wf_<id>.json``) — present on the
   * terminal snapshot so the UI can link to the complete output.
   */
  statePath?: string;
  /**
   * The run's returned ``result.question`` — the actual research question the
   * workflow answered (NOT the workflow's meta description). Terminal snapshot
   * only; ``null`` for runs that ended without a result (killed/aborted).
   */
  resultQuestion?: string | null;
  /** The run's returned ``result.summary`` — its answer. Terminal only. */
  resultSummary?: string | null;
}

/**
 * Decoded ``session.workflow_progress`` SSE frame. ``id`` is the ``Workflow``
 * launch tool_use_id — the frontend keys progress by it so the card attaches
 * to the matching tool call. The host JSON-stringifies ``state`` on the wire;
 * ``parseWorkflowProgress`` re-parses it.
 */
export interface WorkflowProgressEvent {
  id: string;
  run_id: string;
  state: WorkflowState;
  message_id: string;
}

File: cli/internal/cmd/session.go
package cmd

import (
	"context"
	"fmt"

	"github.com/spf13/cobra"

	"code.xiaobangtouzi.com/valuz/valuz-oss/cli/internal/backend"
	errs "code.xiaobangtouzi.com/valuz/valuz-oss/cli/internal/errors"
)

// newSessionCmd builds `valuz session ...` — session lifecycle around the
// product's conversation model (quick chat / project chat share the same
// session primitive). Wraps the sessions API into functional commands.
func newSessionCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "session",
		Short: "Manage sessions (conversations)",
	}
	cmd.AddCommand(
		newSessionCreateCmd(),
		newSessionListCmd(),
		newSessionShowCmd(),
		newSessionInterruptCmd(),
		newSessionEventsCmd(),
		newSessionSendCmd(),
		newSessionApproveCmd(),
	)
	return cmd
}

func newSessionApproveCmd() *cobra.Command {
	var pendingID, decision, message string
	cmd := &cobra.Command{
		Use:   "approve <id> --pending <pending-id> --decision <verb>",
		Short: "Resolve a pending approval (approve|approve_for_session|reject|answer)",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if pendingID == "" {
				return errs.New(errs.KindUsage, "--pending <pending-id> is required")
			}
			switch decision {
			case "approve", "approve_for_session", "reject":
			case "answer":
			case "":
				return errs.New(errs.KindUsage, "--decision is required (approve|approve_for_session|reject|answer)")
			default:
				return errs.New(errs.KindUsage, "unsupported --decision %q", decision)
			}
			opts, err := Options(cmd)
			if err != nil {
				return err
			}
			client := backend.NewControlClient(opts.BackendURL, bearerToken(opts))
			var resp struct {
				SessionID string `json:"session_id"`
				PendingID string `json:"pending_id"`
				Decision  string `json:"decision"`
			}
			body := backend.SessionActionRequest{PendingID: pendingID, Decision: decision, Message: message}
			if err := client.Post(cmd.Context(), "/v1/sessions/"+args[0]+"/actions", body, &resp); err != nil {
				return err
			}
			fmt.Fprintf(cmd.OutOrStdout(), "resolved %s (%s)\n", resp.PendingID, resp.Decision)
			return nil
		},
	}
	f := cmd.Flags()
	f.StringVar(&pendingID, "pending", "", "pending approval id (from requires_action events)")
	f.StringVar(&decision, "decision", "", "approve|approve_for_session|reject|answer")
	f.StringVar(&message, "message", "", "optional message with the decision")
	return cmd
}

func newSessionCreateCmd() *cobra.Command {
	var (
		projectID      string
		cwd            string
		agentSlug      string
		modelID        string
		providerID     string
		runtimeID      string
		permissionMode string
		title          string
		quick          bool
		mcpSlugs       []string
		skillIDs       []string
	)
	cmd := &cobra.Command{
		Use:   "create",
		Short: "Create a session (quick chat by default, or bound to a project)",
		RunE: func(cmd *cobra.Command, _ []string) error {
			opts, err := Options(cmd)
			if err != nil {
				return err
			}
			client := backend.NewControlClient(opts.BackendURL, bearerToken(opts))
			_ = quick // --quick is the default fallback; kept for explicitness

			pid, err := idOrQuick(client, cmd.Context(), projectID, cwd)
			if err != nil {
				return err
			}
			s, err := createSession(client, cmd, pid, sessionCreateOpts{
				Title:          title,
				AgentSlug:      agentSlug,
				ModelID:        modelID,
				ProviderID:     providerID,
				RuntimeID:      runtimeID,
				PermissionMode: permissionMode,
				MCPSlugs:       mcpSlugs,
				SkillIDs:       skillIDs,
			})
			if err != nil {
				return err
			}
			fmt.Fprintf(cmd.OutOrStdout(), "created session %s (project %s, status %s)\n", s.ID, s.ProjectID, s.Status)
			return nil
		},
	}
	f := cmd.Flags()
	f.StringVar(&projectID, "project", "", "project id (default: quick chat)")
	f.StringVar(&cwd, "cwd", "", "resolve project from a local directory")
	f.BoolVar(&quick, "quick", false, "quick chat (default when no project given)")
	f.StringVar(&agentSlug, "agent", "", "agent slug to bind")
	f.StringVar(&modelID, "model", "", "model id")
	f.StringVar(&providerID, "provider", "", "provider id")
	f.StringVar(&runtimeID, "runtime", "", "runtime: claude_agent|codex|deepagents|deepseek_harness")
	f.StringVar(&permissionMode, "permission-mode", "", "default|auto_review|full_access")
	f.StringVar(&title, "title", "", "session title")
	f.StringSliceVar(&mcpSlugs, "mcp", nil, "MCP data source slugs (repeatable)")
	f.StringSliceVar(&skillIDs, "skill", nil, "extra skill ids to attach (repeatable)")
	return cmd
}

func newSessionListCmd() *cobra.Command {
	var projectID string
	cmd := &cobra.Command{
		Use:   "list",
		Short: "List sessions (optionally filtered by project)",
		RunE: func(cmd *cobra.Command, _ []string) error {
			opts, err := Options(cmd)
			if err != nil {
				return err
			}
			client := backend.NewControlClient(opts.BackendURL, bearerToken(opts))
			path := "/v1/sessions"
			if projectID != "" {
				path += "?project_id=" + projectID
			}
			var list backend.SessionListResponse
			if err := client.Get(cmd.Context(), path, &list); err != nil {
				return err
			}
			for _, s := range list.Sessions {
				agent := ""
				if s.AgentSlug != nil {
					agent = *s.AgentSlug
				}
				fmt.Fprintf(cmd.OutOrStdout(), "%-36s  %-10s  %-18s  %s\n", s.ID, s.Status, s.Runtime, agent)
			}
			return nil
		},
	}
	cmd.Flags().StringVar(&projectID, "project", "", "filter by project id")
	return cmd
}

func newSessionShowCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "show <id>",
		Short: "Show a session's detail",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			opts, err := Options(cmd)
			if err != nil {
				return err
			}
			client := backend.NewControlClient(opts.BackendURL, bearerToken(opts))
			var s backend.SessionDetail
			if err := client.Get(cmd.Context(), "/v1/sessions/"+args[0], &s); err != nil {
				return err
			}
			agent := ""
			if s.AgentSlug != nil {
				agent = *s.AgentSlug
			}
			model := ""
			if s.ModelID != nil {
				model = *s.ModelID
			}
			fmt.Fprintf(cmd.OutOrStdout(), "id:             %s\n", s.ID)
			fmt.Fprintf(cmd.OutOrStdout(), "project:        %s\n", s.ProjectID)
			fmt.Fprintf(cmd.OutOrStdout(), "status:         %s\n", s.Status)
			fmt.Fprintf(cmd.OutOrStdout(), "runtime:        %s\n", s.Runtime)
			fmt.Fprintf(cmd.OutOrStdout(), "permission:     %s\n", s.PermissionMode)
			fmt.Fprintf(cmd.OutOrStdout(), "agent:          %s\n", agent)
			fmt.Fprintf(cmd.OutOrStdout(), "model:          %s\n", model)
			return nil
		},
	}
}

func newSessionInterruptCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "interrupt <id>",
		Short: "Interrupt a running session",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			opts, err := Options(cmd)
			if err != nil {
				return err
			}
			client := backend.NewControlClient(opts.BackendURL, bearerToken(opts))
			var s backend.SessionDetail
			if err := client.Post(cmd.Context(), "/v1/sessions/"+args[0]+"/interrupt", nil, &s); err != nil {
				return err
			}
			fmt.Fprintf(cmd.OutOrStdout(), "interrupted session %s (status %s)\n", s.ID, s.Status)
			return nil
		},
	}
}

func newSessionEventsCmd() *cobra.Command {
	var outputFormat string
	var stream bool
	cmd := &cobra.Command{
		Use:   "events <id>",
		Short: "Show session events (durable history, or live stream)",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			opts, err := Options(cmd)
			if err != nil {
				return err
			}
			sessionID := args[0]
			client := backend.NewControlClient(opts.BackendURL, bearerToken(opts))
			streamClient := backend.NewStreamClient(opts.BackendURL, bearerToken(opts))
			streamClient.ReconnectMaxAttempts = 0

			if stream {
				return streamClient.Stream(cmd.Context(), "/v1/sessions/"+sessionID+"/events/stream", 0, func(ctx context.Context, f *backend.SSEFrame) error {
					if f.IsHeartbeat() {
						return nil
					}
					if outputFormat == "jsonl" {
						raw, err := frameJSON(f)
						if err != nil {
							return err
						}
						fmt.Fprintln(cmd.OutOrStdout(), string(raw))
						return nil
					}
					fmt.Fprintf(cmd.OutOrStdout(), "%s %s\n", *f.EventType, payloadSummary(f))
					return nil
				})
			}

			var history struct {
				SessionID string `json:"session_id"`
				Items     []struct {
					Seq   int64 `json:"seq"`
					Event struct {
						EventType string            `json:"event_type"`
						Payload   map[string]string `json:"payload"`
					} `json:"event"`
					Timestamp *int64  `json:"timestamp"`
					EventUID  *string `json:"event_uid"`
				} `json:"items"`
			}
			if err := client.Get(cmd.Context(), "/v1/sessions/"+sessionID+"/events", &history); err != nil {
				return err
			}
			for _, it := range history.Items {
				line := fmt.Sprintf("%d %s", it.Seq, it.Event.EventType)
				if t := it.Event.Payload["text"]; t != "" {
					line += " " + truncate(t, 80)
				}
				fmt.Fprintln(cmd.OutOrStdout(), line)
			}
			return nil
		},
	}
	cmd.Flags().StringVarP(&outputFormat, flagOutput, "o", "", "output format: human|jsonl")
	cmd.Flags().BoolVar(&stream, "stream", false, "subscribe live events (SSE)")
	return cmd
}

func newSessionSendCmd() *cobra.Command {
	var prompt string
	cmd := &cobra.Command{
		Use:   "send <id>",
		Short: "Send a message to a session (non-blocking dispatch)",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if prompt == "" {
				return errs.New(errs.KindUsage, "--prompt is required")
			}
			opts, err := Options(cmd)
			if err != nil {
				return err
			}
			client := backend.NewControlClient(opts.BackendURL, bearerToken(opts))
			var s backend.SessionDetail
			if err := client.Post(cmd.Context(), "/v1/sessions/"+args[0]+"/messages",
				backend.SessionMessageRequest{Prompt: prompt}, &s); err != nil {
				return err
			}
			fmt.Fprintf(cmd.OutOrStdout(), "dispatched (session %s, status %s)\n", s.ID, s.Status)
			return nil
		},
	}
	cmd.Flags().StringVar(&prompt, "prompt", "", "message to send")
	return cmd
}

func payloadSummary(f *backend.SSEFrame) string {
	if t := f.Payload["text"]; t != "" {
		return truncate(t, 80)
	}
	if m := f.Payload["message"]; m != "" {
		return truncate(m, 80)
	}
	return ""
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "…"
}
